In [5]:
import pandas as pd
import numpy as np
import json
from pandas.io.json import json_normalize
import os
import glob

In [6]:
folder_path = '../data/train_timelines/'
file_list = glob.glob(os.path.join(folder_path, '*.json'))

In [7]:
participantid = []
currentgold = []
totalgold = []
minionskilled = []
xp = []
matchid = []
teamid = []
cs_abilityHaste = []
cs_abilityPower = []
cs_armor = []
cs_attackDamage = []
cs_attackSpeed = []
cs_health = []
cs_healthMax = []
dd_totalDamageTaken = []
dd_pDamageDoneToChamps = []
dd_totDamageDone = []
positionX = []
positionY = []



def file_buster(file_path):
    with open(file_path) as json_file:
        data = json.load(json_file)
    matchida= data['matchId']
    for i, j in data['frames'][-1]['participantFrames'].items():  #looks at last frame, looks at participantFrames equiv to last_stats
        participantid.append(i)
        currentgold.append(j['currentGold'])
        totalgold.append(j['totalGold'])
        xp.append(j['xp'])
        minionskilled.append(j['minionsKilled'])
        matchid.append(matchida)
        if int(i) < 6:
            teamid.append('100')
        else:
            teamid.append('200')
        cs_stats = j['championStats']
        cs_abilityHaste.append(cs_stats['abilityHaste'])
        cs_abilityPower.append(cs_stats['abilityPower'])
        cs_armor.append(cs_stats['armor'])
        cs_attackDamage.append(cs_stats['attackDamage'])
        cs_attackSpeed.append(cs_stats['attackSpeed'])
        cs_health.append(cs_stats['health'])
        cs_healthMax.append(cs_stats['healthMax'])
    
        d_stats = j['damageStats']
        dd_totalDamageTaken.append(d_stats['totalDamageTaken'])
        dd_pDamageDoneToChamps.append(d_stats['totalDamageDoneToChampions'])
        dd_totDamageDone.append(d_stats['totalDamageDone'])
    
        pos = j['position']
        positionX.append(pos['x'])
        positionY.append(pos['y']) 

In [8]:
count = 0
for file in file_list:
    #print('loading file: ' + str(count))
    file_buster(file)
    count+=1
print(str(count) + " files loaded")

8000 files loaded


In [9]:
di = {'matchId':matchid, 'teamId':teamid, 'participantId' : participantid, 'currentGold': currentgold, \
      'totalGold':totalgold, 'xp': xp, 'cs_abilityHaste': cs_abilityHaste, 'cs_abilityPower':cs_abilityPower,\
     'cs_armor': cs_armor, 'cs_attackDamage': cs_attackDamage,'cs_attackSpeed':cs_attackSpeed, \
     'cs_health': cs_health, 'cs_healthMax': cs_healthMax, 'dd_totalDamageTaken':dd_totalDamageTaken, \
     'dd_pDamageDoneToChamps' : dd_pDamageDoneToChamps, 'dd_totDamageDone': dd_totDamageDone, \
     'positionX': positionX, 'positionY': positionY}
za = pd.DataFrame(di)
za    

matchId teamId participantId  currentGold  totalGold    xp  \
0            0    100             1          341       2741  4141   
1            0    100             2            5       3045  2861   
2            0    100             3          365       3545  4712   
3            0    100             4         1214       2789  2894   
4            0    100             5          566       2026  2551   
...        ...    ...           ...          ...        ...   ...   
79995      999    200             6         1127       1127   851   
79996      999    200             7          844       1694  1546   
79997      999    200             8           56       1731  2489   
79998      999    200             9          212       1512  1257   
79999      999    200            10          316       1251  1222   

       cs_abilityHaste  cs_abilityPower  cs_armor  cs_attackDamage  \
0                    0                9       112               84   
1                    0              102        53               72   
2                    0              113        56               70   
3                    0                0        51              114   
4                    0               56        44               69   
...                ...              ...       ...              ...   
79995                0                0        43               80   
79996                0                0        51               90   
79997                0               18        47               72   
79998                0                0        42               92   
79999                0               46        36               61   

       cs_attackSpeed  cs_health  cs_healthMax  dd_totalDamageTaken  \
0                 115       1448          1525                 3965   
1                 238       1234          1234                10014   
2                 107          0          1325                 3835   
3                 125       1082          1082                 2445   
4                 106       1102          1102                 1376   
...               ...        ...           ...                  ...   
79995             114        800           800                  527   
79996             106        635           867                 4173   
79997             113          0          1389                 2613   
79998             124        856           856                  959   
79999             104        794           794                  999   

       dd_pDamageDoneToChamps  dd_totDamageDone  positionX  positionY  
0                        3446             22588        394        461  
1                        2471             43236        986       1629  
2                        2700             33594       8102       7583  
3                        2112             21804      13118       2895  
4                        1138              5274      11778       2819  
...                       ...               ...        ...        ...  
79995                     126              2094      14340      14390  
79996                     408             23496       8707      10696  
79997                    1004              8191       7038       8022  
79998                     238              7068      13660       4454  
79999                     742              2319      12404       6126  

[80000 rows x 18 columns]

In [10]:
champion_mastery = pd.read_csv('../data/champion_mastery.csv')
participants_train = pd.read_csv('../data/participants_train.csv')

In [11]:
champion_points = pd.merge(participants_train, champion_mastery, on = ['summonerId', 'championId'], how = 'left' ).drop(['chestGranted', 'tokensEarned'], axis = 1)
champion_points

matchId  teamId  participantId  summonerId  summonerLevel championName  \
0            0     100              1           0            303  Mordekaiser   
1            0     100              2           1            616        Sylas   
2            0     100              3           2            667    Lissandra   
3            0     100              4           3            860      Caitlyn   
4            0     100              5           4            325      Morgana   
...        ...     ...            ...         ...            ...          ...   
79995     7999     200              6       13979            595       Yorick   
79996     7999     200              7       39643             38     Volibear   
79997     7999     200              8        5570            498       Anivia   
79998     7999     200              9       10228            733       Twitch   
79999     7999     200             10        1684            574       Zilean   

       championId  championLevel  championPoints  
0              82            6.0         61890.0  
1             517            5.0         37664.0  
2             127            6.0         40303.0  
3              51            6.0         96304.0  
4              25            7.0        608721.0  
...           ...            ...             ...  
79995          83            6.0         48037.0  
79996         106            1.0           135.0  
79997          34            7.0         49027.0  
79998          29            7.0        174089.0  
79999          26            6.0         30703.0  

[80000 rows x 9 columns]

In [12]:
#convert to int types
coltomergeon = ['matchId', 'teamId', 'participantId']

for x in za[coltomergeon]:
    za[x] = za[x].astype(int)
    
for x in champion_points[coltomergeon]:
    champion_points[x] = champion_points[x].astype(int)

In [13]:
frames_champs = pd.merge(za, champion_points, on = ['matchId', 'teamId', 'participantId'], how = 'left')
frames_champs

matchId  teamId  participantId  currentGold  totalGold    xp  \
0            0     100              1          341       2741  4141   
1            0     100              2            5       3045  2861   
2            0     100              3          365       3545  4712   
3            0     100              4         1214       2789  2894   
4            0     100              5          566       2026  2551   
...        ...     ...            ...          ...        ...   ...   
79995      999     200              6         1127       1127   851   
79996      999     200              7          844       1694  1546   
79997      999     200              8           56       1731  2489   
79998      999     200              9          212       1512  1257   
79999      999     200             10          316       1251  1222   

       cs_abilityHaste  cs_abilityPower  cs_armor  cs_attackDamage  ...  \
0                    0                9       112               84  ...   
1                    0              102        53               72  ...   
2                    0              113        56               70  ...   
3                    0                0        51              114  ...   
4                    0               56        44               69  ...   
...                ...              ...       ...              ...  ...   
79995                0                0        43               80  ...   
79996                0                0        51               90  ...   
79997                0               18        47               72  ...   
79998                0                0        42               92  ...   
79999                0               46        36               61  ...   

       dd_pDamageDoneToChamps  dd_totDamageDone  positionX  positionY  \
0                        3446             22588        394        461   
1                        2471             43236        986       1629   
2                        2700             33594       8102       7583   
3                        2112             21804      13118       2895   
4                        1138              5274      11778       2819   
...                       ...               ...        ...        ...   
79995                     126              2094      14340      14390   
79996                     408             23496       8707      10696   
79997                    1004              8191       7038       8022   
79998                     238              7068      13660       4454   
79999                     742              2319      12404       6126   

       summonerId  summonerLevel  championName  championId  championLevel  \
0               0            303   Mordekaiser          82            6.0   
1               1            616         Sylas         517            5.0   
2               2            667     Lissandra         127            6.0   
3               3            860       Caitlyn          51            6.0   
4               4            325       Morgana          25            7.0   
...           ...            ...           ...         ...            ...   
79995        8653             45       DrMundo          36            1.0   
79996         345            658        Khazix         121            6.0   
79997        5527             95         Sylas         517            7.0   
79998        8654            710          Jinx         222            7.0   
79999        8268            531           Lux          99            7.0   

       championPoints  
0             61890.0  
1             37664.0  
2             40303.0  
3             96304.0  
4            608721.0  
...               ...  
79995           344.0  
79996         33794.0  
79997        163699.0  
79998        114586.0  
79999        486731.0  

[80000 rows x 24 columns]

In [14]:
#bringing in frames_champs_stats to aggregate
frames_champs_stats = pd.read_csv('../data/frames_champs_stats.csv').drop(['Unnamed: 0', 'summonerId'], axis =1)
frames_champs_stats

matchId  teamId  participantId  currentGold  totalGold    xp  \
0            0     100              1          341       2741  4141   
1            0     100              2            5       3045  2861   
2            0     100              3          365       3545  4712   
3            0     100              4         1214       2789  2894   
4            0     100              5          566       2026  2551   
...        ...     ...            ...          ...        ...   ...   
79995      999     200              6         1127       1127   851   
79996      999     200              7          844       1694  1546   
79997      999     200              8           56       1731  2489   
79998      999     200              9          212       1512  1257   
79999      999     200             10          316       1251  1222   

       cs_abilityHaste  cs_abilityPower  cs_armor  cs_attackDamage  ...  \
0                    0                9       112               84  ...   
1                    0              102        53               72  ...   
2                    0              113        56               70  ...   
3                    0                0        51              114  ...   
4                    0               56        44               69  ...   
...                ...              ...       ...              ...  ...   
79995                0                0        43               80  ...   
79996                0                0        51               90  ...   
79997                0               18        47               72  ...   
79998                0                0        42               92  ...   
79999                0               46        36               61  ...   

       positionX  positionY  summonerLevel  championName  championId  \
0            394        461            303   Mordekaiser          82   
1            986       1629            616         Sylas         517   
2           8102       7583            667     Lissandra         127   
3          13118       2895            860       Caitlyn          51   
4          11778       2819            325       Morgana          25   
...          ...        ...            ...           ...         ...   
79995      14340      14390             45       DrMundo          36   
79996       8707      10696            658        Khazix         121   
79997       7038       8022             95         Sylas         517   
79998      13660       4454            710          Jinx         222   
79999      12404       6126            531           Lux          99   

       championLevel  championPoints      Score      winrt       KDA  
0                6.0         61890.0  53.420000  51.098333  2.168333  
1                5.0         37664.0  59.128889  50.896667  2.420000  
2                6.0         40303.0  51.655000  51.915000  2.700000  
3                6.0         96304.0  77.272500  49.610000  2.377500  
4                7.0        608721.0  48.066000  49.992000  2.710000  
...              ...             ...        ...        ...       ...  
79995            1.0           344.0  39.121667  48.556667  2.088333  
79996            6.0         33794.0  49.550000  49.707500  2.745000  
79997            7.0        163699.0  59.128889  50.896667  2.420000  
79998            7.0        114586.0  71.025000  52.655000  2.512500  
79999            7.0        486731.0  57.472500  50.838750  2.905000  

[80000 rows x 26 columns]

In [15]:
results = pd.DataFrame({'matchId': frames_champs_stats['matchId'].unique()})

cols_to_pivot = ['totalGold', 'xp',  'cs_abilityHaste', 'cs_abilityPower','cs_armor', 'cs_attackDamage','cs_attackSpeed', \
     'cs_health', 'cs_healthMax', 'dd_totalDamageTaken', \
     'dd_pDamageDoneToChamps' ,'dd_totDamageDone', \
     'positionX', 'positionY' , 'championName', 'championLevel', 'championPoints' ]


for i in frames_champs_stats[cols_to_pivot]:
    print(i)
    pivot_table = pd.pivot(frames_champs_stats, values = i, index = 'matchId', columns = ['teamId', 'participantId'])
    
    
    pivot_table.columns = [str(i)+str('_'.join(map(str, c))) for c in pivot_table]
    
    results = pd.merge(results, pivot_table, on='matchId', how = 'inner')
    
    
   

totalGold
xp
cs_abilityHaste
cs_abilityPower
cs_armor
cs_attackDamage
cs_attackSpeed
cs_health
cs_healthMax
dd_totalDamageTaken
dd_pDamageDoneToChamps
dd_totDamageDone
positionX
positionY
championName
championLevel
championPoints


In [16]:
results

matchId  totalGold100_1  totalGold100_2  totalGold100_3  totalGold100_4  \
0           0            2741            3045            3545            2789   
1           1            2888            4549            3947            3600   
2          10            2664            3850            3712            4927   
3         100            2814            4072            3289            5535   
4        1000            3164            4548            3811            3998   
...       ...             ...             ...             ...             ...   
7995      995            3368            3486            3951            3682   
7996      996            3842            3779            3663            4833   
7997      997            3583            3761            3037            3093   
7998      998            3000            3458            2832            3846   
7999      999            2473            2139            2022            2184   

      totalGold100_5  totalGold200_6  totalGold200_7  totalGold200_8  \
0               2026            4380            4114            3372   
1               2678            3715            2891            3003   
2               2841            4655            3186            4058   
3               3880            3516            4619            2940   
4               3808            4078            3602            3068   
...              ...             ...             ...             ...   
7995            2280            3232            3053            3835   
7996            2739            3209            3274            3461   
7997            2325            4141            3821            3343   
7998            2462            4427            4389            3161   
7999            1208            1127            1694            1731   

      totalGold200_9  ...  championPoints100_1  championPoints100_2  \
0               3419  ...              61890.0              37664.0   
1               3084  ...              32069.0             890336.0   
2               2968  ...              29595.0             190520.0   
3               2653  ...              53370.0              64717.0   
4               2830  ...             972443.0              99843.0   
...              ...  ...                  ...                  ...   
7995            3174  ...              13401.0               6910.0   
7996            3667  ...              36651.0               2700.0   
7997            4427  ...              73803.0              82340.0   
7998            2770  ...              22451.0               2662.0   
7999            1512  ...             915982.0             623034.0   

      championPoints100_3  championPoints100_4  championPoints100_5  \
0                 40303.0              96304.0             608721.0   
1                605169.0             151187.0             548849.0   
2                 16424.0              67522.0               5179.0   
3               2703244.0             220056.0             639229.0   
4                 20610.0              13271.0              89289.0   
...                   ...                  ...                  ...   
7995              56731.0              17585.0              24890.0   
7996             373159.0              33958.0              20172.0   
7997               8594.0              33273.0                387.0   
7998             103213.0               6158.0             205005.0   
7999               7016.0             224410.0             199242.0   

      championPoints200_6  championPoints200_7  championPoints200_8  \
0               1180052.0              45215.0             118831.0   
1                404369.0             423591.0              16728.0   
2                 81384.0              15555.0             171780.0   
3               1676567.0             236945.0                277.0   
4                 62110.0              88791.0               7079.0   
...                   ...           

In [17]:
summary_aggs = pd.DataFrame({'matchId': frames_champs_stats['matchId'].unique()})

cols_to_agg = ['totalGold', 'xp', 'cs_abilityHaste', 'cs_abilityPower','cs_armor', 'cs_attackDamage','cs_attackSpeed', \
     'cs_health', 'cs_healthMax', 'dd_totalDamageTaken', \
     'dd_pDamageDoneToChamps' ,'dd_totDamageDone', \
     'positionX', 'positionY' ,  'championLevel', 'championPoints', 'Score', 'winrt', 'KDA' ]


for i in frames_champs_stats[cols_to_agg]:
    print(i)
    
    medians = pd.DataFrame(frames_champs_stats.groupby(['matchId', 'teamId'])[i].median()).reset_index()
    medians = medians.pivot(index = 'matchId', columns = 'teamId', values = i).reset_index()   
    medians = medians.rename(columns = {c: 'med'+str(c)+str(i) for c in medians.columns if c not in ['matchId']})
    
    avg = pd.DataFrame(frames_champs_stats.groupby(['matchId', 'teamId'])[i].mean()).reset_index()
    avg = avg.pivot(index = 'matchId', columns = 'teamId', values = i).reset_index()   
    avg = avg.rename(columns = {c: 'avg'+str(c)+str(i) for c in avg.columns if c not in ['matchId']})
    
    skw = pd.DataFrame(frames_champs_stats.groupby(['matchId', 'teamId'])[i].skew()).reset_index()
    skw = skw.pivot(index = 'matchId', columns = 'teamId', values = i).reset_index()   
    skw = skw.rename(columns = {c: 'skw'+str(c)+str(i) for c in skw.columns if c not in ['matchId']})
    
    quant = pd.DataFrame(frames_champs_stats.groupby(['matchId', 'teamId'])[i].quantile(0.75)).reset_index()
    quant = quant.pivot(index = 'matchId', columns = 'teamId', values = i).reset_index()   
    quant = quant.rename(columns = {c: 'quant'+str(c)+str(i) for c in quant.columns if c not in ['matchId']})
    
    summed = pd.DataFrame(frames_champs_stats.groupby(['matchId', 'teamId'])[i].sum()).reset_index()
    summed = summed.pivot(index = 'matchId', columns = 'teamId', values = i).reset_index()   
    summed = summed.rename(columns = {c: 'summed'+str(c)+str(i) for c in summed.columns if c not in ['matchId']})
     
    
    summary_aggs = pd.merge(summary_aggs, medians, on = 'matchId', how = 'inner')
    summary_aggs = pd.merge(summary_aggs, avg, on = 'matchId', how = 'inner')
    summary_aggs = pd.merge(summary_aggs, skw, on = 'matchId', how = 'inner')
    summary_aggs = pd.merge(summary_aggs, quant, on = 'matchId', how = 'inner')
    summary_aggs = pd.merge(summary_aggs, summed, on = 'matchId', how = 'inner')
    print(summary_aggs.shape)
   

totalGold
(8000, 11)
xp
(8000, 21)
cs_abilityHaste
(8000, 31)
cs_abilityPower
(8000, 41)
cs_armor
(8000, 51)
cs_attackDamage
(8000, 61)
cs_attackSpeed
(8000, 71)
cs_health
(8000, 81)
cs_healthMax
(8000, 91)
dd_totalDamageTaken
(8000, 101)
dd_pDamageDoneToChamps
(8000, 111)
dd_totDamageDone
(8000, 121)
positionX
(8000, 131)
positionY
(8000, 141)
championLevel
(8000, 151)
championPoints
(8000, 161)
Score
(8000, 171)
winrt
(8000, 181)
KDA
(8000, 191)


In [18]:
summary_aggs

matchId  med100totalGold  med200totalGold  avg100totalGold  \
0           0           2789.0           3419.0           2829.2   
1           1           3600.0           3003.0           3532.4   
2          10           3712.0           3186.0           3598.8   
3         100           3880.0           3516.0           3918.0   
4        1000           3811.0           3068.0           3865.8   
...       ...              ...              ...              ...   
7995      995           3486.0           3174.0           3353.4   
7996      996           3779.0           3274.0           3771.2   
7997      997           3093.0           3821.0           3159.8   
7998      998           3000.0           3161.0           3119.6   
7999      999           2139.0           1512.0           2005.2   

      avg200totalGold  skw100totalGold  skw200totalGold  quant100totalGold  \
0              3507.8        -0.354389        -0.805480             3045.0   
1              2963.2         0.212950        -0.378273             3947.0   
2              3602.0         0.619625         0.917242             3850.0   
3              3621.6         1.001764         0.130549             4072.0   
4              3135.8        -0.089803        -0.193662             3998.0   
...               ...              ...              ...                ...   
7995           3154.6        -1.571211         0.022237             3682.0   
7996           3355.8         0.097870         0.984773             3842.0   
7997           3708.2        -0.687886        -0.502268             3583.0   
7998           3332.8         0.294833        -0.176405             3458.0   
7999           1463.0        -1.535031        -0.344686             2184.0   

      quant200totalGold  summed100totalGold  ...  med100KDA  med200KDA  \
0                4114.0               14146  ...     2.4200   2.450000   
1                3084.0               17662  ...     2.4425   2.809167   
2                4058.0               17994  ...     2.6250   2.667500   
3                4380.0               19590  ...     2.6175   2.230000   
4                3602.0               19329  ...     2.4000   2.642500   
...                 ...                 ...  ...        ...        ...   
7995             3232.0               16767  ...     2.6925   2.593750   
7996             3461.0               18856  ...     2.5925   2.667500   
7997             4141.0               15799  ...     2.6175   2.458000   
7998             4389.0               15598  ...     2.5500   2.420000   
7999             1694.0               10026  ...     2.3775   2.512500   

      avg100KDA  avg200KDA  skw100KDA  skw200KDA  quant100KDA  quant200KDA  \
0      2.475167   2.555000  -0.193084   0.229854      2.70000     2.885000   
1      2.697500   2.702833   1.704324  -1.308220      2.64250     2.905000   
2      2.518900   2.547000  -1.130095  -1.631006      2.64250     2.710000   
3      2.608500   2.364300  -0.263517   0.400380      2.87250     2.744000   
4      2.486500   2.489750  -0.405390  -0.109445      2.77000     2.945000   
...         ...        ...        ...        ...          ...          ...   
7995   2.717667   2.634750   0.671970   0.174310      2.87250     2.885000   
7996   2.498274   2.427929  -1.781171  -0.252401      2.59375     2.777143   
7997   2.536900   2.282100  -1.394480  -0.320293      2.67500     2.527500   
7998   2.585500   2.434107   0.937167  -0.967676      2.62500     2.704286   
7999   2.644500   2.534167   0.409307  -0.400582      3.18000     2.745000   

      summed100KDA  summed200KDA  
0        12.375833     12.775000  
1        13.487500     13.514167  
2        12.594500     12.735000  
3        13.042500     11.821500  
4        12.432500     12.448750  
...            ...           ...  
7995     13.588333     13.173750  
7996     12.491369     12.139643  
7997     12.684500     11.410500  
7998     12.927500     12.170536  
7999     13.222500     12.670833  

[8000

In [19]:
#list(summary_aggs.columns)

In [20]:
train_winners = pd.read_csv('../data/train_winners.csv')
train_winners

matchId  winner
0           0     200
1           1     200
2           2     200
3           3     100
4           4     200
...       ...     ...
7995     7995     200
7996     7996     100
7997     7997     200
7998     7998     200
7999     7999     100

[8000 rows x 2 columns]

In [21]:
aggregates_df = pd.merge(summary_aggs, train_winners, on ='matchId', how = 'inner')
aggregates_df

matchId  med100totalGold  med200totalGold  avg100totalGold  \
0           0           2789.0           3419.0           2829.2   
1           1           3600.0           3003.0           3532.4   
2          10           3712.0           3186.0           3598.8   
3         100           3880.0           3516.0           3918.0   
4        1000           3811.0           3068.0           3865.8   
...       ...              ...              ...              ...   
7995      995           3486.0           3174.0           3353.4   
7996      996           3779.0           3274.0           3771.2   
7997      997           3093.0           3821.0           3159.8   
7998      998           3000.0           3161.0           3119.6   
7999      999           2139.0           1512.0           2005.2   

      avg200totalGold  skw100totalGold  skw200totalGold  quant100totalGold  \
0              3507.8        -0.354389        -0.805480             3045.0   
1              2963.2         0.212950        -0.378273             3947.0   
2              3602.0         0.619625         0.917242             3850.0   
3              3621.6         1.001764         0.130549             4072.0   
4              3135.8        -0.089803        -0.193662             3998.0   
...               ...              ...              ...                ...   
7995           3154.6        -1.571211         0.022237             3682.0   
7996           3355.8         0.097870         0.984773             3842.0   
7997           3708.2        -0.687886        -0.502268             3583.0   
7998           3332.8         0.294833        -0.176405             3458.0   
7999           1463.0        -1.535031        -0.344686             2184.0   

      quant200totalGold  summed100totalGold  ...  med200KDA  avg100KDA  \
0                4114.0               14146  ...   2.450000   2.475167   
1                3084.0               17662  ...   2.809167   2.697500   
2                4058.0               17994  ...   2.667500   2.518900   
3                4380.0               19590  ...   2.230000   2.608500   
4                3602.0               19329  ...   2.642500   2.486500   
...                 ...                 ...  ...        ...        ...   
7995             3232.0               16767  ...   2.593750   2.717667   
7996             3461.0               18856  ...   2.667500   2.498274   
7997             4141.0               15799  ...   2.458000   2.536900   
7998             4389.0               15598  ...   2.420000   2.585500   
7999             1694.0               10026  ...   2.512500   2.644500   

      avg200KDA  skw100KDA  skw200KDA  quant100KDA  quant200KDA  summed100KDA  \
0      2.555000  -0.193084   0.229854      2.70000     2.885000     12.375833   
1      2.702833   1.704324  -1.308220      2.64250     2.905000     13.487500   
2      2.547000  -1.130095  -1.631006      2.64250     2.710000     12.594500   
3      2.364300  -0.263517   0.400380      2.87250     2.744000     13.042500   
4      2.489750  -0.405390  -0.109445      2.77000     2.945000     12.432500   
...         ...        ...        ...          ...          ...           ...   
7995   2.634750   0.671970   0.174310      2.87250     2.885000     13.588333   
7996   2.427929  -1.781171  -0.252401      2.59375     2.777143     12.491369   
7997   2.282100  -1.394480  -0.320293      2.67500     2.527500     12.684500   
7998   2.434107   0.937167  -0.967676      2.62500     2.704286     12.927500   
7999   2.534167   0.409307  -0.400582      3.18000     2.745000     13.222500   

      summed200KDA  winner  
0        12.775000     200  
1        13.514167     200  
2        12.735000     100  
3        11.821500     100  
4        12.448750     100  
...            ...     ...  
7995     13.173750     100  
7996     12.139643     100  
7997     11.410500     200  
7998     12.170536     200  
7999     12.670833     100  

[8000 rows x 192 columns]

In [17]:
aggregates_df.to_csv('../data/aggregates_df.csv')

# FITTING GRADIENT BOOST

In [22]:
df = aggregates_df.dropna()
df.shape

(8000, 192)

In [23]:
[col for col in df if col.startswith('quant')]


['quant100totalGold',
 'quant200totalGold',
 'quant100xp',
 'quant200xp',
 'quant100cs_abilityHaste',
 'quant200cs_abilityHaste',
 'quant100cs_abilityPower',
 'quant200cs_abilityPower',
 'quant100cs_armor',
 'quant200cs_armor',
 'quant100cs_attackDamage',
 'quant200cs_attackDamage',
 'quant100cs_attackSpeed',
 'quant200cs_attackSpeed',
 'quant100cs_health',
 'quant200cs_health',
 'quant100cs_healthMax',
 'quant200cs_healthMax',
 'quant100dd_totalDamageTaken',
 'quant200dd_totalDamageTaken',
 'quant100dd_pDamageDoneToChamps',
 'quant200dd_pDamageDoneToChamps',
 'quant100dd_totDamageDone',
 'quant200dd_totDamageDone',
 'quant100positionX',
 'quant200positionX',
 'quant100positionY',
 'quant200positionY',
 'quant100championLevel',
 'quant200championLevel',
 'quant100championPoints',
 'quant200championPoints',
 'quant100Score',
 'quant200Score',
 'quant100winrt',
 'quant200winrt',
 'quant100KDA',
 'quant200KDA']

### REMOVE POSITION COLUMNS

In [24]:
positions = [col for col in df if col.endswith('positionX')]
positions

['med100positionX',
 'med200positionX',
 'avg100positionX',
 'avg200positionX',
 'skw100positionX',
 'skw200positionX',
 'quant100positionX',
 'quant200positionX',
 'summed100positionX',
 'summed200positionX']

In [25]:
#df2. remove position
df2 = df.drop(['med100positionX',
 'med200positionX',
 'avg100positionX',
 'avg200positionX',
 'skw100positionX',
 'skw200positionX',
 'summed100positionX',
 'summed200positionX',
 'quant100positionX',
 'quant200positionX',
 'quant100positionY',
 'quant200positionY',
 'med100positionY',
 'med200positionY',
 'avg100positionY',
 'avg200positionY',
 'skw100positionY',
 'skw200positionY',
 'summed100positionY',
 'summed200positionY'            
              ], axis =1)

### REMOVE SUMMED!

In [26]:
df3 = df2.drop([col for col in df2 if col.startswith('summed')], axis = 1)

In [27]:

df4 = df3.drop(df3.filter(like = 'abilityHaste').columns, axis = 1)

In [28]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn import preprocessing
import warnings
warnings.filterwarnings("ignore")

In [60]:
X = df.drop(['winner'], axis = 1)
y = df['winner']
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=38)

In [61]:
gbc=GradientBoostingClassifier(n_estimators=500,learning_rate=0.1,random_state=145,max_features=5 )
gbc.fit(X_train,y_train)

GradientBoostingClassifier(max_features=5, n_estimators=500, random_state=145)

In [62]:
print(confusion_matrix(y_test, gbc.predict(X_test)))

[[589 241]
 [239 531]]


In [63]:
print("GBC accuracy is %2.4f" % accuracy_score( 
     y_test, gbc.predict(X_test)))

GBC accuracy is 0.7000


### on aggregates only, stdev removed -- 70%
### adding champion stats --0.7119
### removing positions --0.7050
### removing abilityHaste --0.7019
### added back abilityHaste, skwPositions, summed, -- 0.7037
### a. remove positions, but have summed --0.6975 
### b. remove positions, remove summed -- 0.6987
### c. keeping everything -- 0.7069
### d. removing positions --0.6925 df2
### e. removing summed --0.7025 df3
### f. removing abilityhaste --0.7037 df4

In [64]:
from sklearn.metrics import classification_report

pred=gbc.predict(X_test)

print(classification_report(y_test, pred, digits = 4))

              precision    recall  f1-score   support

         100     0.7114    0.7096    0.7105       830
         200     0.6878    0.6896    0.6887       770

    accuracy                         0.7000      1600
   macro avg     0.6996    0.6996    0.6996      1600
weighted avg     0.7000    0.7000    0.7000      1600



https://vitalflux.com/ridge-classification-concepts-python-examples/

In [29]:
# RIDGE LOGISTIC REGRESSION

In [38]:
from sklearn.linear_model import RidgeClassifier
from sklearn.preprocessing import StandardScaler

In [65]:
# Create standardized training dataset
#
sc = StandardScaler()
X_train_norm = sc.fit_transform(X_train)
X_test_norm = sc.transform(X_test)
#
# Create RidgeClassifier instance
#
rdgclassifier = RidgeClassifier()
rdgclassifier.fit(X_train_norm, y_train)
#
# Score the classifier
#
rdgclassifier.score(X_test_norm, y_test)

0.70875

In [ ]:
### remove positions, kept abilityHaste, kept summed 0.72
### remove positions, kept abilityHaste, remove summed 0.71875
### keeping everything 0.70875
### removing pos 0.709375
### removing summed 0.709375

# Retrying Gradient Boost added with original features.

In [40]:
originaldf = pd.read_csv('../data/data_201_cols.csv').drop(['winner' , 'championName100_1',  #trying it without categoricals
 'championName100_2',
 'championName100_3',
 'championName100_4',
 'championName100_5',
 'championName200_6',
 'championName200_7',
 'championName200_8',
 'championName200_9',
 'championName200_10' ], 1)

In [66]:
#df has everything
#df2 has pos removed
#df3 has sum removed
#df4 has abilityHaste removed
newdf = pd.merge(originaldf, df, how = 'inner', on = 'matchId').drop(['Unnamed: 0'], 1)

In [67]:
newdf = newdf.fillna(0)

In [68]:
newdf[newdf.columns[newdf.isna().any()].tolist()].isna().sum()

Series([], dtype: float64)

In [44]:
#removing abilityhaste from original df
newdf_ahrem = newdf.drop(newdf.filter(like = 'abilityHaste').columns, axis = 1)
newdf_ahrem.shape

(8000, 310)

In [45]:
#removing position form original df
newdf_posrem = newdf.drop(newdf.filter(like = 'position').columns, axis =1)
newdf_posrem.shape

(8000, 300)

In [69]:
X = newdf.drop(['winner'], axis = 1)
y = newdf['winner']

In [70]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state = 21)

In [71]:
gbc=GradientBoostingClassifier(n_estimators=500,learning_rate=0.05,random_state=100,max_features=5 )
gbc.fit(X_train,y_train)

GradientBoostingClassifier(learning_rate=0.05, max_features=5, n_estimators=500,
                           random_state=100)

In [72]:
print(confusion_matrix(y_test, gbc.predict(X_test)))

[[614 215]
 [250 521]]


In [73]:
print("GBC accuracy is %2.4f" % accuracy_score( 
     y_test, gbc.predict(X_test)))

GBC accuracy is 0.7094


### 73.00 after dropping standard deviations and filling nans with 0's --using df2
### 72.69 after dropping skew and standard deviation --using df3
### 73.12 removing std, adding avg and skew
## 73.44 removing positions, .7388 changing seed
### 72.62 adding back skw positions, adding summed
### 72.88 removing all positions, adding summed
### 71.81 removing positions, removing summed
### a.  0.7131 removing positions, keeping summed
### b.  0.7113 remove positions, remove summed
### c.  0.7094 keeping everything (df)
### d.  0.7106 removing positions (df2)
### e.  0.7137 removing summed (df3)
### f.  0.7169 removing abilityhaste(df4)
### g. 0.7087 removing abilityHaste from newdf (newdfahrem)


In [74]:
from sklearn.metrics import classification_report

pred=gbc.predict(X_test)

print(classification_report(y_test, pred, digits=4))

              precision    recall  f1-score   support

         100     0.7106    0.7407    0.7253       829
         200     0.7079    0.6757    0.6914       771

    accuracy                         0.7094      1600
   macro avg     0.7093    0.7082    0.7084      1600
weighted avg     0.7093    0.7094    0.7090      1600



In [ ]:
#RIDGE FEATURE SELECTION

In [612]:
from sklearn.feature_selection import SequentialFeatureSelector
from time import time
from sklearn.linear_model import RidgeCV


ridge = RidgeCV(alphas=np.logspace(-6, 6, num=5)).fit(X, y)
importance = np.abs(ridge.coef_)
feature_names = np.array(X.columns)


tic_fwd = time()
sfs_forward = SequentialFeatureSelector(
    ridge, n_features_to_select=10, direction="forward"
).fit(X, y)
toc_fwd = time()

# tic_bwd = time()
# sfs_backward = SequentialFeatureSelector(
#     ridge, n_features_to_select=2, direction="backward"
# ).fit(X, y)
# toc_bwd = time()

print(
    "Features selected by forward sequential selection: "
    f"{feature_names[sfs_forward.get_support()]}"
)
print(f"Done in {toc_fwd - tic_fwd:.3f}s")
# print(
#     "Features selected by backward sequential selection: "
#     f"{feature_names[sfs_backward.get_support()]}"
# )
# print(f"Done in {toc_bwd - tic_bwd:.3f}s")

Features selected by forward sequential selection: ['cs_abilityPower200_8' 'cs_healthMax200_7' 'avg100totalGold'
 'avg200totalGold' 'med200championLevel' 'avg100championLevel'
 'avg200championLevel' 'skw100championLevel' 'avg200winrt' 'avg100KDA']
Done in 37.380s


from https://scikit-learn.org/stable/auto_examples/feature_selection/plot_select_from_model_diabetes.html#sphx-glr-auto-examples-feature-selection-plot-select-from-model-diabetes-py

In [ ]:
#GBC FEATURE SELECTION

In [75]:
from sklearn.feature_selection import SequentialFeatureSelector
from time import time
#from sklearn.linear_model import RidgeCV


gbc_mod = GradientBoostingClassifier(n_estimators=100,learning_rate=0.1,random_state=100,max_features=5).fit(X_train, y_train)
#importance = np.abs(ridge.coef_)
feature_names = np.array(X_train.columns)


tic_fwd = time()
sfs_forward = SequentialFeatureSelector(
    gbc_mod, n_features_to_select='auto', tol = 0.001, scoring = 'accuracy', direction="forward", n_jobs=-1
).fit(X_train, y_train)
toc_fwd = time()

# tic_bwd = time()
# sfs_backward = SequentialFeatureSelector(
#     ridge, n_features_to_select=2, direction="backward"
# ).fit(X, y)
# toc_bwd = time()

print(
    "Features selected by forward sequential selection: "
    f"{feature_names[sfs_forward.get_support()]}"
)
print(f"Done in {toc_fwd - tic_fwd:.3f}s")
# print(
#     "Features selected by backward sequential selection: "
#     f"{feature_names[sfs_backward.get_support()]}"
# )
# print(f"Done in {toc_bwd - tic_bwd:.3f}s")

Features selected by forward sequential selection: ['dd_totDamageDone100_4' 'positionY200_8' 'championPoints100_1'
 'championPoints200_9' 'KDA200_10' 'avg100totalGold' 'avg200totalGold'
 'med100cs_abilityHaste' 'avg100positionY']
Done in 3799.985s


In [76]:
features_selected_by_fwd_sel = ['dd_totDamageDone100_4' 'positionY200_8' 'championPoints100_1'
 'championPoints200_9' 'KDA200_10' 'avg100totalGold' 'avg200totalGold'
 'med100cs_abilityHaste' 'avg100positionY']

In [ ]:
X_new = X[]

In [551]:
features=[]
for columns in X.columns:
    features.append(columns)

In [552]:
imp_features = gbc.feature_importances_

df_imp_features = pd.DataFrame({"features":features}).join(pd.DataFrame({"weights":imp_features}))
df_imp_features.sort_values(by=['weights']).head(50)

features   weights
228       quant200cs_attackSpeed  0.000000
21          cs_abilityPower100_1  0.000000
198        med200cs_abilityPower  0.000031
289                  skw100Score  0.000106
33                 cs_armor100_3  0.000116
70               cs_health200_10  0.000150
275        quant100championLevel  0.000151
202        skw200cs_abilityPower  0.000167
24          cs_abilityPower100_4  0.000172
159                   Score200_9  0.000195
152                   Score100_2  0.000200
40                cs_armor200_10  0.000205
276        quant200championLevel  0.000205
166                   winrt200_6  0.000231
42          cs_attackDamage100_2  0.000246
135           championLevel100_5  0.000248
64                cs_health100_4  0.000263
156                   Score200_6  0.000268
160                  Score200_10  0.000299
49          cs_attackDamage200_9  0.000305
26          cs_abilityPower200_6  0.000318
117               positionX200_7  0.000319
27          cs_abilityPower200_7  0.000324
37                 cs_armor200_7  0.000325
197        med100cs_abilityPower  0.000352
151                   Score100_1  0.000366
162                   winrt100_2  0.000383
55           cs_attackSpeed100_5  0.000418
157                   Score200_7  0.000419
30         cs_abilityPower200_10  0.000425
214        med200cs_attackDamage  0.000440
89      dd_totalDamageTaken200_9  0.000448
91   dd_pDamageDoneToChamps100_1  0.000455
174                     KDA100_4  0.000456
57           cs_attackSpeed200_7  0.000457
299                quant100winrt  0.000460
222         med200cs_attackSpeed  0.000478
180                    KDA200_10  0.000479
217        skw100cs_attackDamage  0.000480
45          cs_attackDamage100_5  0.000481
136           championLevel200_6  0.000482
36                 cs_armor200_6  0.000482
112               positionX100_2  0.000484
211             quant100cs_armor  0.000486
134           championLevel100_4  0.000487
179                     KDA200_9  0.000502
80            cs_healthMax200_10  0.000509
194                     skw200xp  0.000515
286                  med200Score  0.000515
226         skw200cs_attackSpeed  0.000522

**TUNING**

In [266]:
from sklearn.model_selection import GridSearchCV

grid = {

    'learning_rate':[0.01,0.05,0.1],

    'n_estimators':np.arange(100,300,100),

}


gb = GradientBoostingClassifier()

gb_cv = GridSearchCV(gb, grid, cv = 4)

gb_cv.fit(X_train,y_train)

print("Best Parameters:",gb_cv.best_params_)

print("Train Score:",gb_cv.best_score_)

print("Test Score:",gb_cv.score(X_test,y_test))

KeyboardInterrupt: 

# Trying Ridge Regression

In [627]:
# Create standardized training dataset
#
sc = StandardScaler()
X_train_norm = sc.fit_transform(X_train)
X_test_norm = sc.transform(X_test)
#
# Create RidgeClassifier instance
#
rdgclassifier = RidgeClassifier()
rdgclassifier.fit(X_train_norm, y_train)
#
# Score the classifier
#
rdgclassifier.score(X_test_norm, y_test)

0.695625

In [ ]:
# REMOVE SKEW from DF2(STDEV REMOVED)

In [112]:
df3 = df2.drop([col for col in df2 if col.startswith('skw')], axis = 1)  #gives .7269


# TRYING NEURAL NETWORK

In [189]:
import seaborn as sns
from sklearn.neural_network import MLPClassifier, MLPRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline

In [553]:
pipe = Pipeline(
    steps = [
        ('scaler', MinMaxScaler()),
        ('nn', MLPClassifier(hidden_layer_sizes = (2,),
                             activation = 'tanh',
                             max_iter = 10000))
    ]
)

In [554]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('nn',
                 MLPClassifier(activation='tanh', hidden_layer_sizes=(2,),
                               max_iter=10000))])

In [555]:
accuracy_score(y_test, pipe.predict(X_test))

0.518125